legion/engine/llri-vk/source.cpp

```diff
@@ -1,14 +1,259 @@
+                        layers.push_back("VK_LAYER_KHRONOS_validation");
+                    break;
+                }
+                case instance_extension_type::GPUValidation:
+                {
+                    if (extension.gpuValidation.enable)
+                    {
+                        enables = {
+                            vk::ValidationFeatureEnableEXT::eGpuAssisted,
+                            vk::ValidationFeatureEnableEXT::eGpuAssistedReserveBindingSlot
+                        };
+                        features = vk::ValidationFeaturesEXT(enables, {});
+                        features.pNext = pNext; //Always apply pNext backwards to simplify optional chaining
+                        pNext = &features;
+                    }
+                    break;
+                }
+                default:
+                {
+                    llri::destroyInstance(result);
+                    return result::ErrorExtensionNotSupported;
+                }
+            }
+        }
+
+        vk::ApplicationInfo appInfo{ desc.applicationName, VK_MAKE_VERSION(0, 0, 0), "Legion::LLRI", VK_MAKE_VERSION(0, 0, 1), VK_HEADER_VERSION_COMPLETE };
+        vk::InstanceCreateInfo ci{ {}, &appInfo, (uint32_t)layers.size(), layers.data(), (uint32_t)extensions.size(), extensions.data() };
+        ci.pNext = pNext;
+
+        vk::Instance vulkanInstance = nullptr;
+        const vk::Result createResult = vk::createInstance(&ci, nullptr, &vulkanInstance);
+
+        if (createResult != vk::Result::eSuccess)
+        {
+            llri::destroyInstance(result);
+            return internal::mapVkResult(createResult);
+        }
+
+        result->m_ptr = vulkanInstance;
+        *instance = result;
+        return result::Success;
     }
 
-    unsigned llri::Context::createSomeId(some_strategy_type creationStrategy) noexcept
+    void destroyInstance(Instance instance)
     {
-        return detail::VKStaticId + static_cast<unsigned>(creationStrategy);
+        if (!instance)
+            return;
+
+        for (auto& [ptr, adapter] : instance->m_cachedAdapters)
+            delete adapter;
+
+        //vk validation layers aren't tangible objects and don't need manual destruction
+
+        delete instance;
+    }
+
+    result InstanceT::enumerateAdapters(std::vector<Adapter>* adapters)
+    {
+        if (adapters == nullptr)
+            return result::ErrorInvalidUsage;
+
+        //Clear internal pointers, lost adapters will have a nullptr internally
+        for (auto& [ptr, adapter] : m_cachedAdapters)
+            adapter->m_ptr = nullptr;
+
+        std::vector<vk::PhysicalDevice> physicalDevices;
+
+        //Get count
+        //Can't use Vulkan.hpp convenience function because we need it to not throw if it fails
+        uint32_t physicalDeviceCount = 0;
+        vk::Result r = static_cast<vk::Instance>(static_cast<VkInstance>(m_ptr)).enumeratePhysicalDevices(&physicalDeviceCount, nullptr);
```

Comment:
intellisense won't help me try to figure this out, but why the double `static_cast`?

Reply:
Because vk::Instance has a conversion operator for VkInstance but not for a raw pointer. VkInstance is an alias for VkInstanceT* so we can go from void* -> VkInstanceT* -> vk::Instance but not directly void* -> vk::Instance